pygmt/base_plotting.py

```diff
@@ -495,8 +495,15 @@ def grdimage(self, grid, **kwargs):
             elif kind == "grid":
                 file_context = lib.virtualfile_from_grid(grid)
             else:
-                raise GMTInvalidInput("Unrecognized data type: {}".format(type(grid)))
-            with file_context as fname:
+                raise GMTInvalidInput(f"Unrecognized data type: {type(grid)}")
+
+            with contextlib.ExitStack() as stack:
+                # shading using an xr.DataArray
+                if "I" in kwargs and data_kind(kwargs["I"]) == "grid":
+                    shading_context = lib.virtualfile_from_grid(kwargs["I"])
+                    kwargs["I"] = stack.enter_context(shading_context)
```

Comment:
Need a unit test for `fig.grdimage(..., shading=grid, ...)` in https://github.com/GenericMappingTools/pygmt/blob/master/pygmt/tests/test_grdimage.py. Try modifying the `grdview` example (see the `drapegrid`):

https://github.com/GenericMappingTools/pygmt/blob/e5e25cf0d85dda9d2b2807258ffef88963a6f9ce/pygmt/tests/test_grdview.py#L236-L247

Reply:
I'll add a unit test for this in a bit and then we can merge it in for PyGMT v0.3.0.